go/stellar/acctbundle/acctbundle.go

```diff
@@ -0,0 +1,673 @@
+
+	return &res, nil
+}
+
+// ErrNoChangeNecessary means that any proposed change to a bundle isn't
+// actually necessary.
+var ErrNoChangeNecessary = errors.New("no account mode change is necessary")
+
+// MakeMobileOnly transforms a stellar1.AccountBundle into a mobile-only
+// bundle.  This advances the revision.  If it's already mobile-only,
+// this function will return ErrNoChangeNecessary.
+func MakeMobileOnly(a *stellar1.BundleRestricted, accountID stellar1.AccountID) error {
+	ws, err := AccountWithSecret(a, accountID)
+	if err != nil {
+		return err
+	}
+	if ws.Mode == stellar1.AccountMode_MOBILE {
+		return ErrNoChangeNecessary
+	}
+	a.Revision++
+	a.Prev = a.OwnHash
+	a.OwnHash = nil
+
+	for i, vis := range a.Accounts {
+		if vis.AccountID == accountID {
+			if vis.Mode == stellar1.AccountMode_MOBILE {
+				return ErrNoChangeNecessary
+			}
+			vis.Mode = stellar1.AccountMode_MOBILE
+			vis.AcctBundleRevision++
+			a.Accounts[i] = vis
+		}
+	}
+
+	return nil
+}
+
+// MakeAllDevices transforms a stellar1.AccountBundle into an all-device
+// bundle.  This advances the revision.  If it's already all-device,
+// this function will return ErrNoChangeNecessary.
+func MakeAllDevices(a *stellar1.BundleRestricted, accountID stellar1.AccountID) error {
+	ws, err := AccountWithSecret(a, accountID)
+	if err != nil {
+		return err
+	}
+	if ws.Mode == stellar1.AccountMode_USER {
+		return ErrNoChangeNecessary
+	}
+	a.Revision++
+	a.Prev = a.OwnHash
+	a.OwnHash = nil
+
+	for i, vis := range a.Accounts {
+		if vis.AccountID == accountID {
+			if vis.Mode == stellar1.AccountMode_USER {
+				return ErrNoChangeNecessary
+			}
+			vis.Mode = stellar1.AccountMode_USER
+			vis.AcctBundleRevision++
+			a.Accounts[i] = vis
+		}
+	}
+
+	return nil
+}
+
+// PukFinder helps this package find puks.
+type PukFinder interface {
+	SeedByGeneration(m libkb.MetaContext, generation keybase1.PerUserKeyGeneration) (libkb.PerUserKeySeed, error)
+}
+
+// DecodeAndUnbox decodes the encrypted and visible encoded bundles and unboxes
+// the encrypted bundle using PukFinder to find the correct puk.  It combines
+// the results into a stellar1.AccountBundle.
+func DecodeAndUnbox(m libkb.MetaContext, finder PukFinder, encodedBundle BundleEncodedB64) (*stellar1.BundleRestricted, stellar1.BundleVersion, error) {
+	encBundle, hash, err := decodeParent(encodedBundle.EncParent)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	puk, err := finder.SeedByGeneration(m, encBundle.Gen)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	parent, parentVersion, err := unboxParent(encBundle, hash, encodedBundle.VisParent, puk)
+	if err != nil {
+		return nil, 0, err
+	}
+	parent.AccountBundles = make(map[stellar1.AccountID]stellar1.AccountBundle)
+	for _, parentEntry := range parent.Accounts {
+		if acctEncB64, ok := encodedBundle.AcctBundles[parentEntry.AccountID]; ok {
+			acctBundle, err := decodeAndUnboxAcctBundle(m, finder, acctEncB64, parentEntry)
+			if err != nil {
+				// XXX keep going?
```

Comment:
xxx
I'd recommend not keep going. If you keep going the client is liable to post another bundle that has even less of whatever got corrupted.